### PR TITLE
Big improvements to Loam-evaluation and graph-validation performance (O(n^2) => O(n))

### PR DESCRIPTION
**Pull request by ClintAtTheBroad**
- Loam evaluation and graph validation now run in O(n) time, down from O(n^2), where n is the number of Tools.
- In practical terms, a pipeline that took ~35 minutes to start executing will now take ~40 seconds.
- Added `BiMap` class, which trades some memory for efficient lookups by both keys and values.
- Tool name mapping in `LoamGraph` is now a `BiMap`, making some operations run in O(1) time that used to run in O(n).  Along with some refactoring to `LoamGraph.withInputStores`, this made loam-evaluation run in O(n) time instead of O(n^2).
- Some operations in `LoamGraphValidation` now use a faster method for diffing sets, and now run in linear instead of quadratic time.
- Added `Sets.fasterSetDiff` for guaranteed-linear set-diffing. Depending on the type of `Set`s being compared, `diff()` might run in linear time or quadratic time, if the Set falls back to the default implementation of `diff` which folds over the left-hand set with `-`.
- `BashScript.escapeString` now has a `StringBuilder`-oriented implementation instead of the previous one based on `String.flatMap`.  The new implementation is much faster (at least 67%) and surprisingly, profiling showed that LS spent a lot of time in `escapeString`.
- Added `--no-validation` CLI flag to optionally disable graph validation.  Without this flag, validation will occur.
- `LoamCompiler` now logs the time various steps took at debug level.

### CODE REVIEW

**Comment by ClintAtTheBroad** [comment]
Note: The integration tests are ongoing.  I'll update this when they pass.

File: src/main/scala/loamstream/loam/LoamGraph.scala [hunk 1]
```diff
@@ -68,7 +69,7 @@ final case class LoamGraph(
     //somewhere else?  For now, following the established pattern.  -Clint Nov 9, 2017
     workDirs: Map[Tool, Path],
     executionEnvironments: Map[Tool, Environment],
-    namedTools: Map[String, Tool]) {
+    namedTools: BiMap[Tool, String]) {
```
Comment (massung) [inline src/main/scala/loamstream/loam/LoamGraph.scala]: As it's a `BiMap` I don't know how much the type ordering matters, but was there an underlying reason to go from `[String, Tool]` to `[Tool, String]` that makes things better?

File: src/main/scala/loamstream/loam/LoamGraph.scala [hunk 2]
```diff
@@ -201,12 +203,20 @@ final case class LoamGraph(
...
 
-    val storeProducersNew = storeProducers.filterNot {
-      case (store, producer) => stores(store) && producer == tool
+    //Remove only at most `stores.size` keys from `storeProducers`.  Previously, this code relied on calling
+    //`storeProducers.filterNot`, which traversed the entire storeProducers Map and ran in O(nStores) time.
+    //The current approach is faster since for non-trivial pipelines, the number of keys to remove from 
+    //storeProducers is usually *much* smaller than the total number of stores, and removing a key from a map is
+    //fast, either O(1) or O(log nStores).
+    val storeProducersKeysToRemove: Iterable[Store] = {
+      stores.map { s => s -> storeProducers.get(s) }.collect { case (s, Some(producer)) if producer == tool => s }
```
Comment (massung) [inline src/main/scala/loamstream/loam/LoamGraph.scala]: I _think_ I grok what the improvement is here... it went from:

```scala
bigListOfThings.filterNot(stores.contains(_))
```

to:

```scala
stores.reduce(bigListOfThings)(a, b => a.remove(b))
```

Where - because of the `BiMap`, `bigListOfThings` is now constant-time lookup.

If that's not the case, then the comment feels a bit off and it took me a bit to get there. But the comment does help. Just let me know if my conclusion was wrong. ;-)

The comment also states things like "is usually *much* smaller"; does this approach break if the number is not much smaller? What would be the worst-case big-O if the graph was not typical? I assume no worse than it currently is, but with words like "usually" it would be nice to get a better overall picture.

File: src/main/scala/loamstream/loam/LoamGraph.scala (hunk 1 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/loam/LoamGraph.scala]: Not really.  This is just a leftover from an incomplete unwinding of a previous change (adding another `Map` going the other direction) that I factored out to `BiMap`.

File: src/main/scala/loamstream/loam/LoamGraph.scala [hunk 3]
```diff
@@ -201,12 +203,20 @@ final case class LoamGraph(
...
     }
+    
+    val storeProducersNew = storeProducers -- storeProducersKeysToRemove
 
-    val storeConsumersNew =
+    val storeConsumersNew = {
       storeConsumers ++ stores.map(store => (store, storeConsumers.getOrElse(store, Set.empty) + tool))
+    }
```
Comment (massung) [inline src/main/scala/loamstream/loam/LoamGraph.scala]: At this point it seems like you compute the producer stores to remove/new and then do the same for consumers? I don't know the typical size of `stores`, but if large, since this is a performance area, maybe worth considering doing all those operations together (`O(n)` as opposed to `c*O(n)`)? 

Given the gains you've already gotten, it's likely not worth it, but it stood out so I thought I'd mention it.

File: src/main/scala/loamstream/loam/LoamGraph.scala [hunk 4]
```diff
@@ -236,18 +246,20 @@ final case class LoamGraph(
       storeConsumers = storeConsumersNew)
   }
 
-  def nameOf(t: Tool): Option[String] = namedTools.collectFirst { case (n, namedTool) if namedTool == t => n }
+  def nameOf(t: Tool): Option[String] = namedTools.get(t)
```
Comment (massung) [inline src/main/scala/loamstream/loam/LoamGraph.scala]: Very nice. 👍

File: src/main/scala/loamstream/loam/LoamGraphValidation.scala [hunk 5]
```diff
@@ -65,47 +66,47 @@ object LoamGraphValidation {
...
-      override def apply(graph: LoamGraph, tool: Tool): Seq[LoamToolIssue[Set[Store]]] = {
-        val missingOutputs = graph.toolOutputs.getOrElse(tool, Set.empty) -- graph.stores
-        issueIf(missingOutputs.nonEmpty,
-          newBulkIssue[Tool, Set[Store]](graph, this, tool, missingOutputs, Severity.Warning,
-            s"The following outputs of tool $tool are missing from the graph: ${missingOutputs.mkString(", ")}"))
-      }
+  object eachToolsOutputStoresArePresent extends LoamToolRule[Set[Store]] {
+    override def apply(graph: LoamGraph, tool: Tool): Seq[LoamToolIssue[Set[Store]]] = {
+      //NB: Use fasterSetDiff because `--` had O(n^2) running time :(
+      val missingOutputs = Sets.fasterSetDiff(graph.toolOutputs.getOrElse(tool, Set.empty), graph.stores)
```
Comment (massung) [inline src/main/scala/loamstream/loam/LoamGraphValidation.scala]: I get nervous about function names including adjectives like `fast`, `slow`, etc. Why would I ever use the "slow" version? If the "fast" version is so much better, why wouldn't we always use it, and then why call it "fast" as opposed to replacing the existing implementation?

So, I guess my questions are: 

* what makes `fasterSetDiff` actually faster?
* is there already something in the standard library that does what `fasterSetDiff` does that wasn't immediately obvious?
* would it make sense to name it something more specific to what it does (and how it's different from `--`)?

File: src/main/scala/loamstream/loam/LoamGraph.scala (hunk 2 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/loam/LoamGraph.scala]: > ```scala
bigListOfThings.filterNot(stores.contains(_))
```
> to:
> ```scala
stores.reduce(bigListOfThings)(a, b => a.remove(b))
```
Kinda, more or less.  The first way traverses a potentially huge Map (it has one entry for every Store) that records what Tool produces which Stores, so it's size is technically unbounded, but in big real-world pipelines is maybe 10s or 100s of thousands.  This traversal builds up a new potentially-huge map that's only smaller by (in practice) ~10s of entries at the most.  This matters because this method gets called (effectively) once for every tool!

File: src/main/scala/loamstream/loam/LoamGraphValidation.scala [hunk 6]
```diff
@@ -140,7 +141,7 @@ object LoamGraphValidation {
     }
   }
 
-  val eachToolHasEitherInputsOrOutputs: LoamToolRule[Unit] = new LoamToolRule[Unit] {
+  object eachToolHasEitherInputsOrOutputs extends LoamToolRule[Unit] {
```
Comment (massung) [inline src/main/scala/loamstream/loam/LoamGraphValidation.scala]: Other than "flavor of the month" coding style (which I totally do and have nothing against), is there a reason to change all these variables to objects? Was there an advantage elsewhere in the PR that I missed?

Note: I'm not against it, I prefer "globals" as singleton objects over `val`, but want to make sure I'm not missing something.

File: src/main/scala/loamstream/util/BashScript.scala [hunk 7]
```diff
@@ -26,10 +28,25 @@ object BashScript extends Loggable {
...
-    case c if charsToBeEscaped(c) => Seq('\\', c)
-    case c => Seq(c)
+  /** 
+   * Escapes string for Bash.
+   * Note new StringBuilder-oriented approach, which is significantly faster than the previous one that relied on
+   * String.flatMap.  Somewhat surprisingly, this method was a significant component of the graph-validation code's
+   * running time before switching to the StringBuilder approach.   
+   */
+  def escapeString(string: String): String = {
+    val builder = new StringBuilder(string.length * 2)
```
Comment (massung) [inline src/main/scala/loamstream/util/BashScript.scala]: Given that this was the case (that escaping strings was done often enough to show up), it might be worth considering making the `StringBuilder` a member of `BashScript` or - better still - a member of a companion object (if it doesn't need to be thread safe) and having it always resident for use, clearing and reusing, might save a lot of time constructing, destroying, GC'ing, etc.

File: src/main/scala/loamstream/util/BiMap.scala [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ * 
+ * Represents a bi-directional one-to-one mapping between As and Bs, with unique keys and values.
+ * Allows efficient lookups (constant-time, assuming the component maps offer constant-time lookups) by 
+ * keys /and/ values.
+ * 
+ * Note: uniqueness of keys and values is guaranteed by overwriting existing mappings, just like with regular 
+ * Maps.  When adding mappings with `+`, this class ensures that both forward and backward mappings
+ * are overwritten if they're present.
+ */
+final case class BiMap[A, B] private (private[util] val forward: Map[A, B], private[util] val backward: Map[B, A]) {
```
Comment (massung) [inline src/main/scala/loamstream/util/BiMap.scala]: Just asking to make 100% sure... unless this is some "special edition" BiMap, the reason bi-directional maps are hard to use is that the data is so rarely a guaranteed 1:1 mapping as opposed to 1:n (the same value may appear multiple times in the map). I just want to make sure this is guaranteed 1:1 and you've taken that into consideration? And that any unit-testing you have in place for it tests against it to make sure it can't happen and have a key overridden by accident that will break the graph?

File: src/main/scala/loamstream/util/Sets.scala [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+   * `--` had O(n^2) running time.  What's more, the default implementation of `diff` for non-HashSets just calls
+   * `--`, hence the need to ensure we have HashSets.  Empirically, using this method makes validation run in linear 
+   * time instead of quadratic time, as it did previously.   
+   */
+  def fasterSetDiff[A](lhs: Set[A], rhs: Set[A]): Set[A] = {
+    import scala.collection.immutable.HashSet
+    
+    def toHashSet(s: Set[A]): HashSet[A] = s match {
+      case hs: HashSet[A] => hs 
+      case _ => s.to[HashSet]
```
Comment (massung) [inline src/main/scala/loamstream/util/Sets.scala]: Given that this is the implementation of `fasterSetDiff`, would it make sense to (over time) switch all the `Set` instances to `HashSet` and get rid of the conversion?

File: src/main/scala/loamstream/loam/LoamGraph.scala (hunk 3 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/loam/LoamGraph.scala]: Profiling didn't turn this bit up, but it might rear its head eventually.  I think the biggest factor here is that the size of `stores` is usually small (think 10s at most).

File: src/main/scala/loamstream/loam/LoamGraph.scala (hunk 4 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/loam/LoamGraph.scala]: Thanks, that felt _very_ good to do.  That `collectFirst` bit was a hack that wasn't a problem until it was.

Comment (massung) [reply src/main/scala/loamstream/loam/LoamGraph.scala]: > ...that wasn't a problem until it was.

Like all code. :-)

File: src/main/scala/loamstream/loam/LoamGraphValidation.scala (hunk 5 above)
Comment (massung) [reply src/main/scala/loamstream/loam/LoamGraphValidation.scala]: After seeing that later the function uses the `HashSet` method, I'd likely rename it to something along those lines: `hashSetDiff` or similar. My 2c.

**Comment by massung** [comment]
Not going to review the tests. I looked at the `BiMap` tests, which are the most important (IMO) for the entire PR. I didn't see a test to guarantee 1:1 mapping (and asserting if something happens that would break it). For example:

```
val m = BiMap[String, Int]("a" -> 1, "b" -> 1)
```

That should absolutely assert. If the test is already there and I missed it, I'm sold on everything else (well, not `fastSetDiff` if you don't mind changing it) unless you wanted to address anything else in particular.

**Comment by massung** [comment]
I'd note one more thing with the `BiMap`: given the 1:1 guaranteed mapping, the `K` and `V` types actually matter as values should always be inserted with `K` and not the other way around. That's why I was curious about the template parameter switch from `[String, Tool]` to `[Tool, String]`. If entries are added to the map using `String` keys, then I'd swap them back to make it obvious and the `BiMap` shouldn't allow inserting by `Tool` (or whatever type you pick for the value).

File: src/main/scala/loamstream/loam/LoamGraphValidation.scala (hunk 6 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/loam/LoamGraphValidation.scala]: Yes, I made these objects so their classes would get meaningful names, instead of `$$anon$15` or whatever.  All those anonymous class names made it so I couldn't find out what part of `LoamGraphValidation` (a big file) was slow when looking at profiler data.  

Once I finished profiling, I just left this alone.  There are downsides/tradeoffs to objects over anonymous classes, namely laziness and more bytecode for lazy initialization, but once this worked, I didn't mess with it further.

File: src/main/scala/loamstream/util/BashScript.scala (hunk 7 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/util/BashScript.scala]: That's a good idea, if memory-management overhead related to the builder was significant.  I have no idea one way or another - this profiling exercise taught me to assume nothing! :)

File: src/main/scala/loamstream/util/BiMap.scala (hunk 8 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/util/BiMap.scala]: Yes, there's a test specifically for that case, and the guard in the constructor provides a basic runtime soundness check.

Comment (massung) [reply src/main/scala/loamstream/util/BiMap.scala]: Hmm... I don't know how big they get, but I'd be concerned with this turning into a performance problem down the road as opposed to a check on insert? 

My thinking is this: 

On insert, it's O(1) to check if the value is already present for a different key. That test is simple. However, in the constructor you're checking *everything* (I'm assuming) to make sure it's OK. That's O(n) for every instance created.

If the `BiMap` is only ever created once and then used to do stuff, then cool; the constructor is the easier place to do that work (and makes the tests easier).

If, however, once created, you're adding to it after-the-fact and creating new `BiMap` instances, then the constructor is likely not a great place.

File: src/main/scala/loamstream/util/Sets.scala (hunk 9 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/util/Sets.scala]: Maybe?  I'm hesitant to do that without a bunch more profiling.  `Set.apply` gives you a variety of `Set` implementations depending on the set size, presumably for good reasons:
```
scala> Set(1).getClass
res2: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.Set$Set1

scala> Set(1,2).getClass
res3: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.Set$Set2

scala> Set(1,2,3).getClass
res5: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.Set$Set3

scala> (1 to 10).toSet.getClass
res6: Class[_ <: scala.collection.immutable.Set[Int]] = class scala.collection.immutable.HashSet$HashTrieSet
```
ie, it uses dumb non-general implementations for small set sizes, like
```scala
class Set2[A] private[collection] (elem1: A, elem2: A) ... {
  def contains(a: A): Boolean = a == elem1 || a == elem2
```
I'm guessing that's faster than some kind of hashtable or trie for small sizes, and it definitely uses less memory.  

TLDR: Switching everything to `HashSet` would add different overheads, and might have unintended consequences.

Comment (massung) [reply src/main/scala/loamstream/util/Sets.scala]: I had no idea. How weird, but probably good for "typical" use cases.

**Comment by ClintAtTheBroad** [comment]
> That should absolutely assert. 

A good catch, thank you.  There were test cases with duplicate keys, but I'd forgotten duplicate values.  That's fixed now.

**Comment by ClintAtTheBroad** [comment]
> I'd note one more thing with the BiMap: given the 1:1 guaranteed mapping, the K and V types actually matter as values should always be inserted with K and not the other way around. That's why I was curious about the template parameter switch from [String, Tool] to [Tool, String]. If entries are added to the map using String keys, then I'd swap them back to make it obvious and the BiMap shouldn't allow inserting by Tool (or whatever type you pick for the value).

`BiMap[K, V]` only has `+` and `++` methods that take `(K, V)`s, not `(V, K)`s.  I guess you could call it "forward-biased".

File: src/main/scala/loamstream/loam/LoamGraphValidation.scala (hunk 5 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/loam/LoamGraphValidation.scala]: > `hashSetDiff`

That sounds good to me.

File: src/main/scala/loamstream/util/BiMap.scala (hunk 8 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/util/BiMap.scala]: > On insert, it's O(1) to check if the value is already present for a different key. That test is simple. However, in the constructor you're checking everything (I'm assuming) to make sure it's OK. That's O(n) for every instance created.

That's something I specifically avoided for that very reason.  The runtime guard is pretty basic - it just checks that the forward and backward maps are the same size, since any difference means the map is unsound, and getting sizes is cheap.  

There are likely (maybe?) ways for a `BiMap` to be unsound with equally-sized forward and backward maps, and for that, the code that adds and removes mappings jumps through some hoops to make sure that duplicate values are removed from both the forward and backward maps in the case any mappings need to be overwritten, and this is unit tested, both for duplicate keys (which was there before) and for duplicate values (which you caught).
